exercises/linked-list/tests/linked-list.rs

```diff
@@ -0,0 +1,242 @@
+    assert_eq!(counter.get(), 0);
+}
+
+
+#[test]
+fn clone_is_equal() {
+    let mut list = LinkedList::new();
+    for i in 0..10 {
+        list.push_back(i);
+    }
+    let mut list2 = clone_list(&list);
+
+    for _ in 0..10 {
+        assert_eq!(list.pop_back(), list2.pop_back());
+    }
+}
+
+#[test]
+fn insert_middle() {
+    let mut list = LinkedList::new();
+    for i in 0..10 {
+        list.push_back(i);
+    }
+    let mut list2 = clone_list(&list);
+
+    {
+        let mut cursor = list.cursor_tail();
+        for _ in 0..4 {
+            cursor.next();
+        }
+        //cursor.seek_forward(4);
+        while let Some(elem) = list2.pop_back() {
+            cursor.insert_after(elem);
+        }
+        //cursor.insert_list_after(&mut list2);
+    }
+
+    assert_eq!(list.len(), 20);
+    assert_eq!(list2.len(), 0);
+
+    let expected = (0..5).chain(0..10).chain(5..10);
+
+    for (exp, &actual) in expected.zip(list.iter()) {
+        assert_eq!(exp, actual);
+    }
+}
+
+#[test]
+fn head_tail_changes_on_push_back() {
+    let mut heads = vec![];
+    let mut tails = vec![];
+    let mut list = LinkedList::new();
+
+    for i in 0..10 {
+        list.push_back(i);
+        heads.push(list.cursor_head().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+        tails.push(list.cursor_tail().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+    }
+    tails.sort();
+    tails.dedup();
+
+    assert_eq!(tails.len(), 1);
+    assert_eq!(tails[0], heads[0]);
+
+    heads.sort();
+    heads.dedup();
+
+    assert_eq!(heads.len(), 10);
+}
+
+#[test]
+fn head_tail_changes_on_push_front() {
+    let mut heads = vec![];
+    let mut tails = vec![];
+    let mut list = LinkedList::new();
+
+    for i in 0..10 {
+        list.push_front(i);
+        heads.push(list.cursor_head().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+        tails.push(list.cursor_tail().peek_mut().map_or(std::ptr::null(), |r| r as *const i32));
+    }
+    heads.sort();
+    heads.dedup();
+
+    assert_eq!(heads.len(), 1);
+    assert_eq!(heads[0], tails[0]);
+
+    tails.sort();
+    tails.dedup();
+
+    assert_eq!(tails.len(), 10);
+}
+
+#[test]
+fn linked_list_is_send_sync() {
+    trait AssertSend: Send {}
+    trait AssertSync: Sync {}
+
+    impl<T: Send> AssertSend for LinkedList<T> {}
+    impl<T: Sync> AssertSync for LinkedList<T> {}
+}
+
+#[allow(dead_code)]
+#[test]
+fn is_covariant() {
+    fn a<'a>(x: LinkedList<&'static str>) -> LinkedList<&'a str> {
+        x
+    }
+}
+
+#[test]
+fn is_generic() {
+    struct Foo;
+    LinkedList::<Foo>::new();
+}
+
+fn clone_list<T: Clone>(list: &LinkedList<T>) -> LinkedList<T> {
+    let mut new_list = LinkedList::new();
+    for element in list.iter().cloned() {
+        new_list.push_back(element);
+    }
+    new_list
+}
+
+// implement some additional functionality using the required interface
+// to limit the amount of code the student has to oversee
+trait ListExt<T> {
```

Comment:
I very much like the idea of using this extension trait to reduce the students' workload. However, we should probably standardize on a single nomenclature pair for the list's ends: either front/back or head/tail. 

I personally conceptualize lists with the head at the front and the tail at the back. It is clear from these implementations that you conceptualize them with the head at the back and the tail at the front. Both conceptualizations are valid, but both will be confusing to people who use the other conceptualization. 

I have a mild (~2/10) preference for using the front/back nomenclature throughout instead of head/tail.

Reply:
You're totally right. I took over the head/tail nomenclature from `std::collections::{LinkedList, VecDeque}` and at times I was somewhat confused which was where. There's no benefit to keeping two names around. I want to keep the `{push,pop}_{back, front}` interfaces consistent with std so I'll start using those everywhere.

Reply:
An alternative to an extension trait is to put these methods and their implementation directly into `impl<T> LinkedList<T>` in `lib.rs`. That way the student can also use them for `FromIterator` or `Extend`

Reply:
That's not a bad idea, though it might be surprising for a student to find some methods pre-implemented for them. If you do this, please add a comment explaining that these methods were trivial to write, but useful, so we implemented them for free to give a leg up.